elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/SQLQueryEngine.java

```diff
@@ -269,22 +290,130 @@ private SQLQuery toPageTotalSQL(SQLQuery sql) {
      * All other code should use this method to generate sql column reference, no matter where the reference is used (
      * select statement, group by clause, where clause, having clause or order by clause).
      *
-     * @param path The predicate path to convert
-     * @param dictionary dictionary to expand joinTo path
+     * @param path the full join path to be resolved.
+     * @param dictionary dictionary instance
      * @return A SQL fragment that references a database column
      */
     public static String generateColumnReference(Path path, EntityDictionary dictionary) {
+        return generateColumnReference(path, new LinkedHashSet<>(), new HashMap<>(), dictionary);
+    }
+
+    /**
+     * Converts a filter predicate path into a SQL column reference.
+     * All other code should use this method to generate sql column reference, no matter where the reference is used (
+     * select statement, group by clause, where clause, having clause or order by clause).
+     *
+     * @param path the full join path to be resolved.
+     * @param toResolve join paths and join paths fragments that are not resolved yet.
+     * @param resolved join paths and join fragments that are already resolved
+     * @param dictionary dictionary instance
+     * @return A SQL fragment that references a database column
+     */
+    public static String generateColumnReference(Path path,
+                                                 LinkedHashSet<Path> toResolve,
```

Comment:
I think we should break this up.  JoinTo should be expanded independently from the expression logic.  Each column that has a JoinTo clause - can create and store a singular path that is fully expanded.  The logic which expands this can check for cycles.

Separately from the JoinTo expansion, we can handle expression expansion.  This logic should be consolidated for metrics and dimensions.   This might require some refactoring of those respective classes.  Let's get together and map out a design.